sales + receivings item attributes support (#68)

jekkos opened the pull request:
After letting this code float for more than a year I finally managed to push through and polish what was there to finish a functional whole. A bit of a sad thing I got demotivated as the scope of the initial design was badly incepted by me and some things needed time to come to fruition. Also I found myself losing a lot of time in designing a good way to manage the attributes in the item form.

I wanted to get this out as I'm leaving on a hike trip for a week to northern Sweden (lapland). I hope that some of you can find the time to review and test the branch before I return so I can do the necessary changes and fixes.

Let me describe briefly what is in here.

### Functionally
* 'Custom' fields and all involved ugliness was removed from the config section
* Item attributes can be created / edited  through the attributes module in the backoffice
* 'Groups' of attributes can be defined. This serves as an easy way to add multiple attributes to an item when creating or updating it. For eg. in my case for running shoes we have both color and size that can be linked to the 'Shoes' Group attribute definition
* Two levels of attribute visibility, being 'sales' and 'receivings'. This means that the values for those attribute definitions will be visible in the respective receipts
* Three 'types' of attribute definitions
 - **DATE**: Renders the field in the item dialogue using a datepicker, and saves the values accordingly. Useful for saving 'expiry dates' on items or the like
 - **DROPDOWN**: Choose from a fixed set of values which you define in the attributes module. Rendered using a select box
 - **TEXT**: Just add some free text as an attribute value. This control is rendered using autocomplete widget and will optimally reuse existing database entries
* Attribute 'values' are searchable through the existing attribute search flag in items screen. This should work as before with custom fields

### Refactoring
* Did some minor refactoring / cleanup here and there, mostly around the datepicker widget and table management

### Extra focus for testing
* Migration script is foreseen to move existing custom fields to the new attribute module system. I might **need some help from others** testing the specific SQL here as I don't have a database that used custom fields, so most testing was done on a clean situation

### Design considerations
I found that I lost a lot of time trying to think of an easy and concise way doing CRUD for attribute values for both the insert and update situation. In the end I found a nice way around by leveraging a server side ajax view refresh cycle that sends out form controls using javascript input. Personally I think this is  cleaner than trying to build the DOM form javascript itself, which is done in quite a lot of other places in the app. 

### Open ends
* One thing that is still missing is a way to show attribute values in the items table by default. Initially I added a visibility flag for this but there's no code yet to handle showing attribute values in the items table.

And now I gotta go make my backpack..

jekkos commented:
@objecttothis thanks for commenting. 
* The branch will be deployed on our demo server again in a couple of minutes, so you can have a quick try there
* We could indeed consider moving some fields to this section, however, will need to go through them to see which ones would be worth moving
* A similar mechanism might be useful adding form controls to customers module (which was asked for some time ago)
* And indeed the DATETIME attribute type uses the user's locale setting for formatting / parsing

odiea commented:
Migration does not work.
Configuration Error 500.
Cannot delete an or modify attribute.
All custom tables even empty tables (updated to attribute tables) show in Items after upgrade.
Let the games begin.

jekkos commented:
@odiea these errors occur after migration then?

odiea commented:
I had to manually add the new attributes. Then I tried to update or delete the current attributes and nothing happened. I don't know if Config worked before adding the attributes. I also did not enable error reporting so I will do that and let you know if any errors show when trying to enter the config module.

odiea commented:
Missing a comma on line 301 and 302 has a comma at the end of the line. Not anything major.
ERROR - 2018-08-17 16:42:48 --> Severity: error --> Exception: syntax error, unexpected ''multi_pack_enabled'' (T_CONSTANT_ENCAPSED_STRING), expecting ')' C:\wamp64\www\attributes\application\controllers\Config.php 302

odiea commented:
Has receipt issues also.
![2018-08-17 11_13_00-wedidit _ powered by ospos 3 3 0](https://user-images.githubusercontent.com/10962177/44279453-a8cc4000-a20e-11e8-93de-01b29ed1d1a9.png)
Receipt quantity changed to 9 after printing receipt.
![2018-08-17 11_13_17-wedidit _ powered by ospos 3 3 0](https://user-images.githubusercontent.com/10962177/44279462-b1247b00-a20e-11e8-89f7-d206c778be55.png)
Another odd receipt. Quantity sold was 1.
![2018-08-17 11_08_33-wedidit _ powered by ospos 3 3 0](https://user-images.githubusercontent.com/10962177/44279487-c7cad200-a20e-11e8-93ab-b7d25332f381.png)

mckaygerhard commented:
seems there's a re-sending on a part of the screen, by 8x

odiea commented:
Here is what is shown after all the attributes are removed from the Item.
![2018-08-17 12_31_12-wedidit _ powered by ospos 3 3 0](https://user-images.githubusercontent.com/10962177/44282858-a58a8180-a219-11e8-836e-26ec8f0060e1.png)

jekkos commented:
Config section is fixed. Indeed I also notice some strange behavior in sales register (adding an item multiple times when linking attributes), but will check that further somewhere next week alas.

jekkos commented:
thanks for the quick feedback though!

jekkos commented:
@odiea Ok no sorry I couldn't stop so I fixed the receipt bug as well. Probably the other issues you're facing are due to the migration doing something wrong. I'll have to check that (if I can move from an old database using custom fields to a new one, to reproduce this). Will take a little more time to set up an old version first and then migrate (as I develop  using docker and thus recreate db everytime)

jekkos commented:
But so that one I will postpone until earliest 27/08 (when I get back from my trip)

odiea commented:
The migration will probably be different for everyone depending on how many Categories or how the custom fields were used.

SteveIreland commented:
I noticed that the field category was dropped from the items table, but it wasn't dropped from the form.  When I go into item maintenance it crashes on the missing field.  Was it meant to be replaced by attributes? If so I'll remove it from my copy of form.php and continue.  Or is there some magic that's supposed to  happen via javascript that didn't happen for me. 

I'm testing from scratch and not running any migrations.

daN4cat commented:
Is categories an attribute now?

odiea commented:
Here is what I see in database.sql. category is still there.
![2018-08-18 04_47_54-c__wamp64_www_attributes_database_database sql - notepad](https://user-images.githubusercontent.com/10962177/44298527-fa71da80-a2a1-11e8-8582-f4f460be21dd.png)

SteveIreland commented:
Okay, I must have built it incorrectly.  I'm also encountering a javascript issue that I'm sure no one else has.  I'll begin anew.

daN4cat commented:
Ok, because I remember we discussed that point and my position was to leave category and have separate concept of attributes group.

SteveIreland commented:
Much better.  Sorry for the extraneous noise.

SteveIreland commented:
The attributes need to be selected for each individual item added.  This is nice because now we can have attributes that are specific to a type of item.  

It would be nice, however, to be able to flag an attribute as auto-selected for all items.  

It would also be nice to set a flag that determines whether or not it is required.

I played around with this and used as my test case the harmonized system code that is needed for India GST.  It is an item attribute, and only has real value in an India GST configuration.  Legally it should be entered for all items since the type of item is used to establish tax rates and the government is probably validating the HSN code against the tax rates used.

The HSN code has to appear on the invoice but since it is simply a numeric value it will need a field label of some type.  It doesn't work if it is just appended to the item name.  So in this case I would need to deselect it as "visible" for sales and receiving and then modify the invoice and receiving document to add it manually so that I can assign a  field label to it.  Right?

For fields that drive program logic there isn't a way to make them "system fields" so that they are automatically included and cannot be removed by the user.  I prefer to avoid user defined fields when program logic is dependent on them but I vaguely recall asking about this a year ago and I thought it was going to be included.  I only raise it for discussion because my intent is to continue with the tax category id as part of the item table and not as part of the dynamic item attributes and I wanted to make sure that was okay with the team.

SteveIreland commented:
A 10 day vacation ... geesh ... I start feeling "work deprived" after 3 days.  Surely they have internet access in Lapland.   :-)

daN4cat commented:
I had two weeks of holidays and I didn't think about work at all.

daN4cat commented on application/views/attributes/item.php:
```diff
@@ -50,27 +49,27 @@
         $("input[name*='attribute_links']").change(function() {
             var definition_id = $(this).data('definition-id');
-            $("input[name='attribute_ids[" + definition_id + "]").val("");
+            $("input[name='attribute_ids[" + definition_id + "]']").val('');
```
I changed this line because it didn't look right to me (you would expect [ [ ] ], but I had not the chance to test it yet.

daN4cat commented on application/views/attributes/form.php:
```diff
@@ -0,0 +1,174 @@
+		{
+			$(form).ajaxSubmit({
+				beforeSerialize: function($form, options) {
+				    $("select[disabled='disabled'").removeAttr('disabled');
```
I'm not sure this line is correct because there is a [ but no ]. Probably it should be ']".
If somebody could change and test this and the other change I made we could get confidence it's fine.

daN4cat commented:
I made a couple of changes to js in views/forms related to attribute (item.php and form.php). I believe they were missing a closing ].
I don't have time to test, also because I would need to change my upgrade to an old version database script to fit my custom fields to attributes.
Somehow I can't get the Docker to work properly on my laptop, something I need to investigate.

So if somebody could give test please.

odiea commented:
"The URI you submitted has disallowed characters". is the error I get when I try to delete an attribute from the attribute table.

SteveIreland commented:
It's building a URL  of `POST /public/[object%20Object]/delete`

odiea commented:
I am getting that error when ever I try to delete anything. Items,Sales,Expenses and so on.

daN4cat commented:
Was it like that before my change?

odiea commented:
I will re-download the original and let you know. I did note that I could not delete an attribute from the table but I did not test other things to delete.

odiea commented:
https://bintray.com/jekkos/opensourcepos/download_file?file_path=opensourcepos.20180817002410.item-attributes.ac44d9.zip. This was the 1st one I tested and yes the issue is in there.

odiea commented:
I just installed the current master and it has no such issues.

SteveIreland commented:
Oh no ... a change in the deep forests of OSPOS where only the wizards dare tread. 

I think the problem may have been introduced in a change to manage_tables.js

WebShells commented:
Hola... is there anyone else having problem loading detailed sales report on demo server ? And i think we should display attributes in Summary items report and detailed sales report O.o ! Any thoughts ?

WebShells commented:
Actually all specific reports ( Employees/Discounts/Customers ) and Transactions... Unable to handle request...

jekkos commented:
Ok so a couple of things for me to check

(-) attribute delete. Did some changes in manage tables that could be the culprit
(-) detailed sales / receivings reports. I agree that we should see the attribute value as columns if they are flagged visible.

objecttothis commented:
@jrkkos, one thing to consider is adding a attribute type of number with customizable precision (whole, 1, 2, 3 places, etc). For example our bookstore will need to add an attribute for item weight which should be restricted to a number, but the number attribute type could later be used to convert many different already existing database fields to attributes.

daN4cat commented:
Welcome back @objecttothis you disappeared for a while. End of the summer?

objecttothis commented:
The building our bookstore owned was being torn down so I needed to focus my attention of not getting ripped off, moving to a new building and getting the government to give us the necessary documents to reopen. As it stands we are having to take the government to court just to force them to give us documents that they want bribes for. At least I can start giving some attention back to OSPOS changes and writing the connector for the CMS we are using and OSPOS to talk to it.

jekkos commented:
Ok fixed some more bugs, attributes added to sales / receivings reports. If anyone can find more bugs, let me know, otherwise it'll be rebase time soon

daN4cat commented:
@jekkos CI is failing because of an issue with cert altnames, maybe you want to check that issue.

daN4cat commented:
If I find the time I'll test, I fixed a couple of small editorial things not worth commenting.
I guess Items view should have a means to display the attributes in the table, what do you think?

daN4cat commented:
BTW, I cannot search for an attribute content in the Items table. Let say I want to search for all the colour attribute white, it doesn't work

daN4cat commented:
Also it might nice to have the possibility to set a default group so you don't need to always select group when you create an Item.

WebShells commented:
@daN4cat you can search attributes by selecting it in search fields, like we used to do for custom fields... Search by attributes, then it will show.. If i got you right...
Loading the attributes in items list is good idea too...

daN4cat commented:
@WebShells indeed. Thanks

WebShells commented:
My pleasure man...

daN4cat commented:
However, the search by Category doesn't seem to work. Even if I select Search Attribute doesn't work (but I would have not expect to work).

daN4cat commented:
I tried deleting an item from the table but got a js error: _Uncaught TypeError: b.fn.bootstrapTable.defaults.formatConfirmDelete is not a function_

daN4cat commented:
Just noticed that search seems broken in general, even by Supplier or simply Item title.

erastus commented:
@jekkos  it would be good to add quantity to the group of attributes

WebShells commented:
I guess it will be better if we add stock number next to multiple attribute to each item.
For example we want to receive 3 blue shirts and 3 red ones small in size we can make them as one item with several Attributes...

erastus commented:
yes, I want to have inventory stock by color and size

jekkos commented:
@daN4cat Ok thanks for review and testing. Attribute search should indeed work if you select it in the dropdown first. I did test that but maybe didn't check regular search. Will have a look tonight.

Also I siltl need to verify how the migration behaves as @odiea reported some issues there. 
And yes I agree on the fact that we might need a new visibility flag to show an attribute in the items table. But maybe i prefer to add it in followup PR and finish up the rough edges here first, as it's quite a big one already.

Also regarding item creation, you can easily keep an attribute selected at creation time if you use the new button instead of regular submit. In that case the popup will stay open if you're entering multiple new items into the system.

jekkos commented:
- Codefactor checks seem pretty useless (all checks beside one stating that Item->save() has become  too complex. Otherwise it's detecting duplicate code in the language files
- Fixed the item search
- No more attributes created after migration if they weren't used before
- Enabled the visibility flag for attributes column in Items table. However I'm not very happy with the result, as the table is easily oversized when adding one attribute (a horizontal scrollbar is added). If anyone has some UI ideas here let me know

daN4cat commented:
I'll disable code factor over the weekend

jekkos commented:
Ok I have removed the webhook now. It could still be useful but then we need to take the time to configure it properly.

jekkos commented:
Latest build of this branch is on DEMO now.

jekkos commented:
Ok one last call for some tests.. then we're ready for merge?

erastus commented:
I'm going to do a test

SteveIreland commented:
I also downloaded a fresh copy and will be testing it this evening

SteveIreland commented:
*Date Attribute Type*
 
The date type doesn't default to the current date and instead defaults to something in the year 2009.

It also appears to be a date-time stamp instead of a simple date field.

Nor am I able to directly type into it.  The only way to change it is to scroll to and select a date ... and that is a long way to scroll.

jekkos commented:
@daN4cat I'm in the process of upgrading my store's install to this version, it has a considerable big database, sale data for > 5 years and > 20k products. So finally moving away from this custom version I  did for storing categories + item sizes, great opportunity to validate performance on the attributes

@SteveIreland Indeed noticed this issue as well, will have a look today or tomorrow.

daN4cat commented:
@jekkos sounds good. I wanted to see if I can spend some time today on cashup to test and see what else I need to add to that.

daN4cat commented:
sorry @jekkos , there is a small conflict to fix since I committed the cashup module.
It's in the autoload.php, just readd attributes at the end. I'm not touching your branch as you may be working on it.

daN4cat commented on application/migrations/sqlscripts/attributes.sql:
```diff
@@ -0,0 +1,185 @@
+ ADD CONSTRAINT `ospos_attribute_links_ibfk_5` FOREIGN KEY (`sale_id`) REFERENCES `ospos_sales` (`sale_id`);
+
+
+UPDATE `ospos_modules` SET `sort` = 120 WHERE `name_lang_key` = 'module_config';
```
I set this to be 900 so it will be always at the end before the office one.
You can remove this.

daN4cat commented on application/migrations/sqlscripts/attributes.sql:
```diff
@@ -0,0 +1,185 @@
+UPDATE `ospos_modules` SET `sort` = 120 WHERE `name_lang_key` = 'module_config';
+
+INSERT INTO `ospos_modules` (`name_lang_key`, `desc_lang_key`, `sort`, `module_id`) VALUES
+ ('module_attributes', 'module_attributes_desc', 110, 'attributes');
```
Better to set it to 107, before the expenses and cashup modules as absolute order.

daN4cat commented on application/migrations/20180905100000_attributes.php:
```diff
@@ -0,0 +1,20 @@
+<?php if (!defined('BASEPATH')) exit('No direct script access allowed');
```
change the date of the model to be 20180910100000, although it would be better to add all of this to the migration to 3.3.0 to be consistent with the previous versions.

daN4cat commented on application/migrations/sqlscripts/attributes.sql:
```diff
@@ -0,0 +1,185 @@
+ `deleted` TINYINT(1) NOT NULL DEFAULT 0,
+ PRIMARY KEY (`definition_id`),
+ KEY `definition_fk` (`definition_fk`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```
The collate, sorry.

daN4cat commented on application/migrations/sqlscripts/attributes.sql:
```diff
@@ -0,0 +1,185 @@
+ `attribute_value` VARCHAR(45) NULL,
+ `attribute_datetime` DATETIME NULL,
+ PRIMARY KEY (`attribute_id`)
+) ENGINE = InnoDB DEFAULT CHARSET=utf8;
```
The collate, sorry.

daN4cat commented on application/migrations/sqlscripts/attributes.sql:
```diff
@@ -0,0 +1,185 @@
+ KEY `sale_id` (`sale_id`),
+ KEY `receiving_id` (`receiving_id`),
+ UNIQUE `attribute_links_uq1` (`attribute_id`, `definition_id`, `item_id`, `sale_id`, `receiving_id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```
The collate, sorry.

daN4cat commented:
My bad, @jekkos I meant the collate not the charset itself.

odiea commented:
It appears that the en-us and en-gb Attribute module lang is missing.

daN4cat commented on application/language/en/attributes_lang.php:
```diff
@@ -0,0 +1,25 @@
+<?php 
```
This module is not required

daN4cat commented on application/language/fr/module_lang.php:
```diff
@@ -1,5 +1,7 @@
 <?php 
 
+$lang["module_attributes"] = "";
```
This is missing in en_GB and en_US

odiea commented:
$lang["module_attributes"] = "Attributes";
$lang["module_attributes_desc"] = "Add, Update, Delete, and Search attributes.";

daN4cat commented:
Fixed `en_GB` and `en_US` and also removed the `en` subfolder.

odiea commented:
I just downloaded this from the Repository about 20 minutes ago.
Odd looking tables in the attributes edition. In the Items table the new columns are not accurate as per the update form
![2018-09-10 17_33_34-](https://user-images.githubusercontent.com/10962177/45330430-a8c81300-b522-11e8-9799-3f8bf4512b6c.png)

In the transactions report the tables are also not valid. You can the the Color has both of the attributes.
![2018-09-10 17_34_15-window](https://user-images.githubusercontent.com/10962177/45330465-e0cf5600-b522-11e8-8d2e-0468f9792f36.png)

daN4cat commented:
@jekkos have you checked the bug reported above?
Also wondering if you managed to do some load testing and checked the performance.

jekkos commented:
Yes I have it on my todo list, but got blocked early this week due to a broken laptop (someone on the train spilled his coffee over it, and I've been trying to get it back to work since a week, after motherboard replace, etc)

SteveIreland commented:
Ack ack ack!!!

Did they ever find his body after you threw him off the train?

erastus commented:
I can not add, the attributes overlap and you can not click
![screenshot_18](https://user-images.githubusercontent.com/6712397/45846453-6d310400-bced-11e8-82a3-3d5cc7b9ac3c.png)

erastus commented:
show attributes in the search
![screenshot_19](https://user-images.githubusercontent.com/6712397/45846983-0876a900-bcef-11e8-9389-32ae50c046a5.png)

erastus commented:
problems when trying to add 2 attributes, when it's a new item.
when you selected one, the other disappears.
![screenshot_20](https://user-images.githubusercontent.com/6712397/45850393-99538180-bcfb-11e8-8f9e-56990cad7317.png)
![screenshot_21](https://user-images.githubusercontent.com/6712397/45850398-9ce70880-bcfb-11e8-9544-ee1f335dd326.png)

jekkos commented:
@erastus thanks for testing, I'll go over those cases this weekend.

jekkos commented:
@erastus I fixed the css issue and the refresh issue as well. However adding the attribute names in the search suggestion is not an easy change. If I understood well a user can configure the field in the suggestion dropdown. So in this case we'd need to query attributes to enable them to show in this config. Feels more like a new feature in one way or another.

In the end it's rather esthetic as you already know on which terms you were querying. The only differentiator could be another attribute added to the suggestion ..

jekkos commented:
@odiea also the refresh for items screen should be ok now. I need to do some kind of pivot operation on the attribute result set (as rows have to become columns, resulset then needs to be appended to the one of the items) and along the way some extra records can be present (because were joining inventory table as well). This should be fixed now. Also the attribute value positions should be correct wrt their column names (when using multiple visible ones)

erastus commented:
thank you, I'll compare it tonight

erastus commented:
does not work, when the items has no attributes or when it only has one attribute.
![screenshot_24](https://user-images.githubusercontent.com/6712397/45930061-13fce680-bf20-11e8-8989-f7862c358e70.png)

jekkos commented:
> does not work, when the items has no attributes or when it only has one attribute.
> ![screenshot_24](https://user-images.githubusercontent.com/6712397/45930061-13fce680-bf20-11e8-8989-f7862c358e70.png)

Just tried locally and couldn't reproduce this case. You mean nothing shows if you have no attrbutes at all? Seems to work fine here.

odiea commented:
This pertains to the Item add or update form.
1. The table shows it being updated but the new Attributes don't show until the rows are refreshed.
2. Now when updating a form the Attributes default to the way they were entered into the Attributes list and not the way that they were entered into the Item form.

erastus commented:
@jekkos. add elements without attributes, does not show results.

jekkos commented:
@erastus , @odiea think I have both fixed now

erastus commented:
@jekkos there are still problems.
![screenshot_27](https://user-images.githubusercontent.com/6712397/46023962-70d8d800-c0ab-11e8-98b0-05217def5732.png)

odiea commented:
With error logging and development implemented the following shows.
The Modal on the one does not show any error.
![2018-09-25 09_51_55-window](https://user-images.githubusercontent.com/10962177/46026557-e98a6500-c0a8-11e8-90dc-9ea1256635b9.png)

jekkos commented:
@odiea @erastus I can't reproduce the error you were talking about.

jekkos commented:
Could you try to test again on an empty install? perhaps on the demo server?

jekkos commented:
installed latest there so test away :dagger:

odiea commented:
With your last commits in Production mode all works great now. But in Development mode the Items table wont load with this error showing. Message also says Undefined offset:7

A PHP Error was encountered

Severity: Notice

Message: Undefined offset: 6

Filename: helpers/tabular_helper.php

Line Number: 423

Backtrace:

File: C:\wamp64\www\itematta\application\helpers\tabular_helper.php
Line: 423
Function: _error_handler

File: C:\wamp64\www\itematta\application\controllers\Items.php
Line: 73
Function: get_item_data_row

File: C:\wamp64\www\itematta\public\index.php
Line: 315
Function: require_once

odiea commented:
That is fixed in development mode but when I try to search for an attribute I get that response and this is also what is showing under parameters.
![2018-09-29 09_13_46-window](https://user-images.githubusercontent.com/10962177/46247420-72b2dc00-c3c8-11e8-8e4f-b8330464a7a3.png)

odiea commented:
I think you nailed it with that change. Great job.

odiea commented:
Like eating an Elephant. A bite at a time. This shows in development mode but appears ok in Production Mode when adding a new item. This happens when selecting an attribute for the Item. 
![2018-09-29 09_41_27-window](https://user-images.githubusercontent.com/10962177/46247687-44370000-c3cc-11e8-81d6-e61b64385aeb.png)

odiea commented:
Fixed. I will download the bintray version in a bit and do more testing. Looks usable now.

odiea commented:
Still has issues in development mode. After you add a new Item in Production mode.
![2018-09-29 10_36_44-window](https://user-images.githubusercontent.com/10962177/46248222-f1614680-c3d3-11e8-9466-9b96cdaabc16.png)

This is Development mode,
![2018-09-29 10_37_43-window](https://user-images.githubusercontent.com/10962177/46248227-0b9b2480-c3d4-11e8-8708-b51dc0d5adbc.png)

jekkos commented:
Ok should be fixed now

odiea commented:
Detailed Receivings and Transactions show the following. It appears it is missing the first attribute but not the attribute column.
![2018-09-29 13_28_49-window](https://user-images.githubusercontent.com/10962177/46249763-e9ad9c00-c3eb-11e8-928c-53c34c0c7892.png)
Here is what is set on the Item.
![2018-09-29 13_27_51-window](https://user-images.githubusercontent.com/10962177/46249770-fcc06c00-c3eb-11e8-8d36-c46a99ac64f8.png)

erastus commented:
now it looks good thanks. @jekkos 
![screenshot_6](https://user-images.githubusercontent.com/6712397/46262166-0072f200-c4c3-11e8-93e3-ccac22909af1.png)

I do not fount by attributes. this happens when you do a search, includes search by attributes.
![screenshot_7](https://user-images.githubusercontent.com/6712397/46262174-04067900-c4c3-11e8-97ec-c5182b3b8b75.png)

jekkos commented:
Ok so still two open points for detailed reports. @erastus searching for attributes only works when you select the flag in items table

jekkos commented:
Refactored the attribute parsing code a bit (reused in both reports and items table) and did some more testing. They should show fine now. There are a couple of unallowed chars in attribute values for which I added validators in the forms. (| and :)

erastus commented:
when you have more than 20 attributes? will all be shown in the table of details items?

jekkos commented:
Just rebased again with master + aligend the code for attribute expansion in reports + items table. @odiea the category field being absent was probably due to a migration that didn't run (just had this error here as well,logout and login resolved it)

jekkos commented:
@odiea seemed to be a introduced bug after the supplier branch was merged. Should be fixed now.

@erastus yes currently there is no limit on number of attributes in the table. it's up to the user.

odiea commented:
Here is your new code. If I comment out $this->db->select('category'); the update works just fine. Is there a reason some of this was added other then the needed attribute stuff? Yes and now there is no longer an issue with the attributes in Transactions showing properly.
```
public function get_multiple_info($item_ids, $location_id)
	{
		$this->db->select('items.*');
		$this->db->select('company_name');
		$this->db->select('GROUP_CONCAT(DISTINCT CONCAT_WS(\':\', definition_id, attribute_value) ORDER BY definition_id SEPARATOR \'|\') AS attribute_values');
		$this->db->select('category');
		$this->db->select('quantity');
		$this->db->from('items');
		$this->db->join('suppliers', 'suppliers.person_id = items.supplier_id', 'left');
		$this->db->join('item_quantities', 'item_quantities.item_id = items.item_id', 'left');
		$this->db->join('attribute_links', 'attribute_links.item_id = items.item_id AND sale_id IS NULL AND receiving_id IS NULL', 'left');
		$this->db->join('attribute_values', 'attribute_links.attribute_id = attribute_values.attribute_id', 'left');
		$this->db->where('location_id', $location_id);
		$this->db->where_in('items.item_id', $item_ids);
		$this->db->group_by('items.item_id');

		return $this->db->get();
	}
```
and here is the existing code.
```
public function get_multiple_info($item_ids, $location_id)
	{
		$this->db->from('items');
		$this->db->join('suppliers', 'suppliers.person_id = items.supplier_id', 'left');
		$this->db->join('item_quantities', 'item_quantities.item_id = items.item_id', 'left');
		$this->db->where('location_id', $location_id);
		$this->db->where_in('items.item_id', $item_ids);

		return $this->db->get();
	}
,,`

odiea commented:
@jekkos If you rebased this would you please recheck the commit that showed the attributes on the Items form it is not showing the stored attributes properly.

jekkos commented:
Ok hope this solves those two last ones..

odiea commented:
It appears to have fixed the last 2 issues. Could use more testing by different users now.

jekkos commented:
@odiea in that case I would suggest to proceed and merge to master. As far as we can see it's stable enough and merging it will make it available to more users that then can report back if needed.
@daN4cat waiting your last approval here

jekkos commented:
Ok it's been sitting here for long enough. Time to pull the trigger.

daN4cat commented:
And what about the performance with big dataset, have you tested it before merging?
I'm really worried about that because when I tested in the past it was slow.

odiea commented:
When an Attributes is set to text the modal does not drop down far enough to delete Visibility attributes.
The right hand scroll bar does nothing.
![2018-10-15 14_47_59-window](https://user-images.githubusercontent.com/10962177/46977545-b0328d00-d089-11e8-85a0-86b431108f15.png)

jekkos commented:
Also item search uses server side pagination which should be pretty fast no?

jekkos commented:
And if there are issues then I'll tackle them as they are reported.

jekkos commented:
@odiea I'll check soon. Might be better to crete new issues now if there are bugs as this thread has become too long

odiea commented:
@jekkos  This may not be an issue after all. I tried it in another browser and it works fine.
